doc/developer/design/20220203_connectors.md

```diff
@@ -0,0 +1,211 @@
+# Connectors
+
+## Summary
+
+This document proposes a design for **connectors**, a new type of catalog object
+that allows common configuration parameters to be shared across sources and
+sinks.
+
+## Overview
+
+Many users of Materialize create families of sources and sinks that share many
+configuration parameters. The current design of `CREATE SOURCE` and `CREATE
+SINK` make this a verbose affair. The problem is particularly acute with
+Avro-formatted Kafka sources that configure authentication:
+
+```sql
+CREATE SOURCE kafka1
+FROM KAFKA BROKER 'kafka:9092' TOPIC 'top1' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password',
+)
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+
+CREATE SOURCE kafka2
+FROM KAFKA BROKER 'kafka:9092' TOPIC 'top2' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password'
+)
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+```
+
+These two source definition differ only in their topic specification, but
+must duplicate eight other configuration parameters.
+
+With the connectors proposed in this document, the `CREATE SOURCE` can instead
+share all the relevant configuration:
+
+```sql
+CREATE CONNECTOR kafka FOR
+KAFKA BROKER 'kafka:9092' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password'
+);
+
+CREATE CONNECTOR schema_registry FOR
+CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+
+CREATE SOURCE kafka1
+FROM KAFKA CONNECTOR kafka TOPIC 'top1'
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTOR schema_registry;
+
+CREATE SOURCE kafka2
+FROM KAFKA CONNECTOR kafka TOPIC 'top2'
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTOR schema_registry;
+```
+
+## Design
+
+### SQL syntax and semantics
+
+Connector create syntax will come in type specific variants and will not include `with_options`, instead completely specifying the potential options in each variant. 
+
+Using connectors syntactically means replacing the source specific keyword such as BROKER or CONNECTION with CONNECTOR and supplying one or more connector names. 
+
+Properties will be merged by applying values in the supplied order with the last write winning, this supports a usage pattern of supplying base values and then progressively more specific overrides while remaining easy to reason about.  
```

Comment:
I initially thought that was the right path, however I realized that to actually punt on that means connectors must specify all options required since combining them with manually specified options is at least as complex as stacking them. 

I am ok with requiring a connector to specify all of the options needed for a source, it just initially felt at odds with the convenience factor that I understood to be one of the main motivations of the feature.

Reply:
Sorry, I don't follow! It seems strictly easier to me to cleave the existing set of options in two, with an option belonging exclusively to either `CREATE CONNECTOR` or `CREATE SOURCE`?